no-invalid-this false flags usage in ES2015 static methods

**What version are you using?**

1.10.3

**What did you do?**

Use `this` in a static class method, like this:

```js
'use strict';

/* eslint-env node, es6 */
/* eslint no-invalid-this: 2 */

class Foo
{
    static foo()
    {
        console.log('foo');
        this.bar();
    }

    static bar()
    {
        console.log('bar');
    }
}

const f = new Foo();
f.constructor.foo();

// console output
foo
bar
```

**What happened?**

When I run `eslint --no-eslintrc` on this code, I get this error:

```text
  11:9  error  Unexpected `this`  no-invalid-this
```

**What did you expect to happen?**

There shouldn't be an error, this code works fine. Using `this` in a static function to refer to the class is valid in node, Safari and Chrome.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29035476-no-invalid-this-false-flags-usage-in-es2015-static-methods?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/docs/rules/no-invalid-this.md
+++ b/docs/rules/no-invalid-this.md
@@ -74,13 +74,6 @@ var obj = {
     }
 };
 
-class Foo {
-    static foo() {
-        this.a = 0;      /*error Unexpected `this`.*/
-        baz(() => this); /*error Unexpected `this`.*/
-    }
-}
-
 foo.forEach(function() {
     this.a = 0;          /*error Unexpected `this`.*/
     baz(() => this);     /*error Unexpected `this`.*/
@@ -170,6 +163,12 @@ Foo.prototype.foo = function foo() {
 class Foo {
     foo() {
         // OK, this is in a method.
+        this.a = 0;
+        baz(() => this);
+    }
+
+    static foo() {
+        // OK, this is in a method (static methods also have valid this).
         this.a = 0;
         baz(() => this);
     }

--- a/lib/ast-utils.js
+++ b/lib/ast-utils.js
@@ -345,7 +345,7 @@ module.exports = {
                 //   class A { set foo() { ... } }
                 //   class A { static foo() { ... } }
                 case "MethodDefinition":
-                    return parent.static;
+                    return false;
 
                 // e.g.
                 //   var foo = function foo() { ... }.bind(obj);

--- a/tests/lib/rules/no-eval.js
+++ b/tests/lib/rules/no-eval.js
@@ -42,6 +42,7 @@ ruleTester.run("no-eval", rule, {
         "var obj = {foo: function() { this.eval('foo'); }}",
         "var obj = {}; obj.foo = function() { this.eval('foo'); }",
         { code: "class A { foo() { this.eval(); } }", parserOptions: { ecmaVersion: 6 } },
+        { code: "class A { static foo() { this.eval(); } }", parserOptions: { ecmaVersion: 6 } },
 
         // Allows indirect eval
         { code: "(0, eval)('foo')", options: [{"allowIndirect": true}] },

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -202,8 +202,8 @@ var patterns = [
         code: "class A {static foo() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
-        valid: [],
-        invalid: [NORMAL, USE_STRICT, MODULES]
+        valid: [NORMAL, USE_STRICT, MODULES],
+        invalid: []
     },
 
     // Constructors.

